### PR TITLE
Fix TransitionReverseCost usage

### PR DESCRIPTION
**Pull request by merkispavel**
# Issue

https://github.com/valhalla/valhalla/issues/2524
https://github.com/valhalla/valhalla/issues/2524#issuecomment-889152247

The original issue is that in some cases main route cost can be greater than alternative cost. This happens because of the bug in the `TransitionCostReverse` usage, i.e we pass parameters of the wrong edge. So what happens is
1. Main route gets less cost before recosting(i.e when connections are found)
2. After recosting main route gets **greater** cost because `TransitionCost` function is used now for cost calculation
image

The picture after the fix
Screenshot 2021-08-11 at 10 28 01

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/thor/bidirectional_astar.cc [hunk 1]
```diff
@@ -230,17 +230,26 @@ inline bool BidirectionalAStar::ExpandInner(baldr::GraphReader& graphreader,
...
+                         ? costing_->EdgeCost(meta.edge, tile, time_info.second_of_week, flow_sources)
+                         : costing_->EdgeCost(opp_edge, t2, time_info.second_of_week, flow_sources));
+
+  // Separate out transition cost.
+  InternalTurn internal_turn;
+  sif::Cost transition_cost;
+  if (FORWARD) {
+    transition_cost = costing_->TransitionCost(meta.edge, nodeinfo, pred);
+  } else {
+    internal_turn = costing_->TurnType(meta.edge->localedgeidx(), nodeinfo, opp_edge, opp_pred_edge);
```
Comment (merkispavel) [inline src/thor/bidirectional_astar.cc:246]: `TurnType` method doesn't look like a trivial operation so I decided to memoize its result and reuse later when adding new label

File: test/gurka/test_recost.cc [hunk 2]
```diff
@@ -7,6 +7,57 @@ using namespace valhalla;
 
 const std::unordered_map<std::string, std::string> build_config{{"mjolnir.shortcuts", "false"}};
 
+TEST(recosting, forward_vs_reverse) {
```
Comment (merkispavel) [inline test/gurka/test_recost.cc:10]: I tried to write a gurka test for bidir-a* for a few hours but it's too tough, sorry

**Comment by merkispavel** [comment]
I RAD tested the changes. Though I didn't expect the diff to be big enough - it is(7%).
- There are good examples where dive bombs got fixed with this PR(see pictures below)
- However, there are cases in the opposite direction

I believe whatever diff we have - we should merge this fix as it makes the routes consistent with costs
image
image
image
image

And a few bad cases from the diff
image
image
image

**Comment by genadz** [comment]
@merkispavel
1. do you have a chance to measure performance diff ?
2. according to the penultimate example, looks like we allow or forbid the turn depending on the expansion order. Did you investigate this example in more detail ?

File: test/gurka/test_recost.cc (hunk 2 above)
Comment (merkispavel) [reply test/gurka/test_recost.cc:10]: correct. it's red on master and green with after the fix

**Review by genadz — APPROVED** [review]

**Comment by merkispavel** [comment]
> 1. do you have a chance to measure performance diff ?
I'm not sure results from mac os will be sane. I had problems with getting stable benchmark results in the past with OS X

> 2. according to the penultimate example, looks like we allow or forbid the turn depending on the expansion order. Did you investigate this example in more detail ?
I looked into a few divebombs ^ a bit deeply. There's a chance that hierarchy limits are the cause: whenever origin point is at the low level road - divebomb is there

**Pull request merged**